File: backend/app/schemas/pid.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.common import EntityBase, SoftDeleteFields


PidNodeCategory = Literal["main", "instrument", "external"]
PidEdgeType = Literal["process", "signal", "control", "electric"]


class PidViewport(BaseModel):
    x: float
    y: float
    zoom: float = Field(gt=0)


class PidNodePosition(BaseModel):
    x: float
    y: float


class PidNode(BaseModel):
    id: str = Field(min_length=1)
    type: str = Field(min_length=1)
    category: PidNodeCategory
    symbolKey: str = Field(min_length=1)
    label: str = Field(default="", max_length=255)
    tag: str = Field(default="", max_length=100)
    position: PidNodePosition
    sourceRef: dict | None = None
    properties: dict = Field(default_factory=dict)


class PidEdge(BaseModel):
    id: str = Field(min_length=1)
    source: str = Field(min_length=1)
    target: str = Field(min_length=1)
    edgeType: PidEdgeType
    label: str = Field(default="", max_length=255)
    style: dict = Field(default_factory=dict)


class PidDiagramPayload(BaseModel):
    processId: int
    version: int = 1
    updatedAt: datetime
    viewport: PidViewport
    nodes: list[PidNode] = Field(default_factory=list)
    edges: list[PidEdge] = Field(default_factory=list)


class PidDiagramOut(PidDiagramPayload):
    pass


class PidProcessOut(EntityBase, SoftDeleteFields):
    location_id: int
    name: str
    description: str | None = None


class PidProcessCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)


class PidProcessUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)

File: frontend/src/types/pid.ts
export type PidNodeCategory = "main" | "instrument" | "external";
export type PidEdgeType = "process" | "signal" | "control" | "electric";
export type PidSymbolSource = "library" | "upload";
export type PidSymbolStandard = "ISA-5.1" | "ISO-14617";

export type PidSymbol = {
  source: PidSymbolSource;
  libraryKey?: string;
  assetUrl?: string;
  standard: PidSymbolStandard;
};

export type PidNodeType = "equipment" | "instrument" | "external";

export type PidInstrumentType =
  | "TT"
  | "PT"
  | "FT"
  | "LT"
  | "ST"
  | "VT"
  | "WT"
  | "IT"
  | "ET"
  | "ZT"
  | "PIC"
  | "TIC"
  | "FIC"
  | "LIC";

export type PidSourceRef = {
  source: "main-equipment" | "field-equipment" | "equipment-in-operation" | "palette";
  id?: number;
  name?: string;
  meta?: { shapeKey?: string; pidSymbol?: PidSymbol | null };
};

export type PidProperties = {
  plc?: string;
  ranges?: string;
  signalType?: string;
  params?: string;
  meta?: Record<string, string | number | boolean | null | undefined>;
};

export type PidNode = {
  id: string;
  type: PidNodeType;
  category: PidNodeCategory;
  symbolKey: string;
  label: string;
  tag: string;
  position: { x: number; y: number };
  sourceRef?: PidSourceRef | null;
  properties: PidProperties;
};

export type PidEdge = {
  id: string;
  source: string;
  target: string;
  edgeType: PidEdgeType;
  label: string;
  style?: Record<string, string | number>;
};

export type PidDiagram = {
  processId: number;
  version: 1;
  updatedAt: string;
  viewport: { x: number; y: number; zoom: number };
  nodes: PidNode[];
  edges: PidEdge[];
};

export type PidProcess = {
  id: number;
  location_id: number;
  name: string;
  description?: string | null;
  is_deleted: boolean;
  deleted_at?: string | null;
  created_at: string;
  updated_at: string;
};
